core/exchange_controller.go

```diff
@@ -3,61 +3,118 @@ package core
 	"github.com/dominant-strategies/go-quai/core/types"
+	"github.com/dominant-strategies/go-quai/log"
 	"github.com/dominant-strategies/go-quai/params"
 )
 
-// CalculateExchangeRate takes in the parent block and the etxs generated in the
-// current block and calculates the new exchange rate
-func CalculateExchangeRate(hc *HeaderChain, block *types.WorkObject, newTokenChoiceSet types.TokenChoiceSet) (*big.Int, *big.Float, *big.Float, error) {
+func CalculateBetaFromMiningChoiceAndConversions(hc *HeaderChain, block *types.WorkObject, parentExchangeRate *big.Int, newTokenChoiceSet types.TokenChoiceSet) (*big.Int, error) {
 
-	// Read the parents beta values
-	betas := rawdb.ReadBetas(hc.headerDb, block.Hash())
-	if betas == nil {
-		return nil, nil, nil, errors.New("could not find the betas stored for parent hash")
+	// Until there are tokenChoicesSetSize of miner token choices, the exchange rate is unchanged
+	if block.NumberU64(common.PRIME_CTX) < params.ControllerKickInBlock+types.C_tokenChoiceSetSize {
+		return params.ExchangeRate, nil
 	}
 
-	if block.NumberU64(common.ZONE_CTX) < types.C_tokenChoiceSetSize {
-		return params.ExchangeRate, betas.Beta0(), betas.Beta1(), nil
+	var totalQiChoices uint64 = 0
+	var totalQuaiChoices uint64 = 0
+
+	tokenChoicesSet := make([]types.TokenChoices, 0)
+	for _, tokenChoices := range newTokenChoiceSet {
+		totalQuaiChoices += tokenChoices.Quai
+		totalQiChoices += tokenChoices.Qi
+		tokenChoicesSet = append(tokenChoicesSet, tokenChoices)
+
+		if tokenChoices.Diff.Cmp(common.Big0) == 0 {
+			panic("token choices diff is zero for the first time")
+		}
+
+		log.Global.Error("token choices set", tokenChoices)
 	}
 
-	// Do the regression to calculate new betas
-	diff, tokenChoice := SerializeTokenChoiceSet(newTokenChoiceSet)
-	var exchangeRate *big.Int
+	sort.Slice(tokenChoicesSet, func(i, j int) bool {
+		return tokenChoicesSet[i].Diff.Cmp(tokenChoicesSet[j].Diff) < 0
+	})
 
-	if len(tokenChoice) != 0 {
-		r := logistic.NewLogisticRegression(betas.Beta0(), betas.Beta1())
-		// If parent is genesis, there is nothing to train
-		exchangeRate = misc.CalculateKQuai(block, r.BigBeta0(), r.BigBeta1())
+	bestScore := uint64(0)
+	bestDiff := big.NewInt(0)
 
-		r.Train(diff, tokenChoice)
+	var left_zeros, right_zeros, left_ones, right_ones uint64 = 0, 0, 0, 0
+	// once the tokenchoices set is sorted the difficulty values are looked from
+	// smallest to the largest. The goal of this algorithm is to find the difficulty point
+	// at which the number of choices of Qi and Quai are equal
+	for i, choice := range tokenChoicesSet {
+		left_zeros = left_zeros + choice.Quai
+		right_zeros = totalQuaiChoices - left_zeros
 
-		return exchangeRate, r.Beta0(), r.Beta1(), nil
-	} else {
-		return block.ExchangeRate(), betas.Beta0(), betas.Beta1(), nil
+		left_ones = left_ones + choice.Qi
+		right_ones = totalQiChoices - left_ones
+
+		score := left_zeros - right_zeros + right_ones - left_ones
```

Comment:
Can you guarantee that this will never be negative?

Reply:
good point, changed the type to int, since it can negative, we should operate in int